packages/tests/e2e/auth.spec.ts

```diff
@@ -60,4 +61,142 @@ test.describe("Authentication", () => {
 
     await expect(formPage.successState).toBeVisible();
   });
+
+  test("Submit incorrect OTP", async ({ page }) => {
+    const testInbox = createTestInbox();
+    const formPage = new FormPage(page);
+
+    await formPage.load();
+
+    await formPage.selectAuthMethod("otp_via_email");
+    await formPage.enterEmail(testInbox.email);
+    await formPage.submitInitialForm();
+
+    const email = await testInbox.getLatestEmail();
+    expect(email).toBeDefined();
+    if (!email) return;
+
+    const incorrectOtp = "000000";
+
+    await formPage.submitOTP(incorrectOtp);
+
+    await expect(formPage.errorMessage).toBeVisible();
+
+    const otp = await testInbox.getOTP(email);
+    expect(otp).toBeDefined();
+    if (!otp) return;
+
+    await formPage.submitOTP(otp);
+    await expect(formPage.errorMessage).not.toBeVisible();
+
+    await expect(formPage.successState).toBeVisible();
+  });
+
+  test("Resend OTP code", async ({ page }) => {
+    const testInbox = createTestInbox();
+    const formPage = new FormPage(page);
+
+    await formPage.load();
+
+    await formPage.selectAuthMethod("otp_via_email");
+    await formPage.enterEmail(testInbox.email);
+    await formPage.submitInitialForm();
+
+    const email = await testInbox.getLatestEmail();
+    expect(email).toBeDefined();
+    if (!email) return;
+
+    const initialOtp = await testInbox.getOTP(email);
+    expect(initialOtp).toBeDefined();
+    if (!initialOtp) return;
+
+    const beforeResend = new Date(Date.now());
+    await formPage.resendOTPCode();
+    await expect(formPage.emailIcon).toBeVisible();
+
+    const maxRetries = 4;
+    let retries = 0;
+    let newOtp = "";
+
+    while (retries < maxRetries) {
```

Comment:
If I understand this right, this code block polls the email service until the email with the new OTP code arrives. This seems to be on a lower level of abstraction when compared to the rest of the steps in this test case. How do you feel about extracting this to a function named like `fetchNewOTP` or something?

Reply:
Would you have it as a method on `testInbox` then?

Reply:
I think it's very specific to the OTP flow, so I would define that function as close as possible to the actual test case.